### PR TITLE
[SPARK-12771][SQL] Simplify CaseWhen code generation

### PR DESCRIPTION
**Pull request by rxin**
The generated code for CaseWhen uses a control variable "got" to make sure we do not evaluate more branches once a branch is true. Changing that to generate just simple "if / else" would be slightly more efficient.

This closes #10737.

### CODE REVIEW

**Comment by rxin** [comment]
@viirya I took a stab at this.

**Comment by viirya** [comment]
Ok.

**Comment by rxin** [comment]
@viirya  want to review this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 1]
```diff
@@ -137,45 +137,55 @@ case class CaseWhen(branches: Seq[(Expression, Expression)], elseValue: Option[E
...
+          ${ev.value} = ${res.value};
         }
       """
-    }.mkString("\n")
+    }
 
-    val elseCase = {
-      if (elseValue.isDefined) {
-        val res = elseValue.get.gen(ctx)
+    var generatedCode = cases.mkString("", "\nelse {\n", "\nelse {\n")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: This will always generate the last else block?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: Yes but that doesn't matter does it? It is just an empty block.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: Yes.

**Comment by viirya** [comment]
LGTM

**Comment by rxin** [comment]
Thanks - going to merge this.

**Pull request closed without merging**